storage-operator/pkg/controller/volume/volume_controller.go

```diff
@@ -88,85 +498,217 @@ type ReconcileVolume struct {
+		reqLogger.Info("pending volume cannot be finalized: requeue")
+		return delayedRequeue(nil)
+	}
 
-	// Set Volume instance as the owner and controller
-	if err := controllerutil.SetControllerReference(instance, pv, r.scheme); err != nil {
-		return reconcile.Result{}, err
+	// Check if a PV is associated to the volume.
+	pv, err := self.getPersistentVolume(ctx, volume)
+	if err != nil {
+		reqLogger.Error(
+			err, "error while looking for backing PersistentVolume: requeue",
+			"PersistentVolume.Name", volume.Name,
+		)
+		return delayedRequeue(err)
 	}
 
-	// Check if this PV already exists
-	found := &corev1.PersistentVolume{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "", Name: pv.Name}, found)
-	if err != nil && errors.IsNotFound(err) {
-		reqLogger.Info("Creating a new PersistentVolume", "PersistentVolume.Name", pv.Name)
-		err = r.client.Create(context.TODO(), pv)
-		if err != nil {
-			return reconcile.Result{}, err
+	// If we have a backing PV we delete it (the finalizer will keep it alive).
+	if pv != nil && pv.GetDeletionTimestamp().IsZero() {
+		if err := self.client.Delete(ctx, pv); err != nil {
+			reqLogger.Error(
+				err, "cannot delete PersistentVolume: requeue",
+				"PersistentVolume.Name", volume.Name,
+			)
+			return delayedRequeue(err)
 		}
-
-		// PV created successfully - don't requeue
-		return reconcile.Result{}, nil
-	} else if err != nil {
-		return reconcile.Result{}, err
+		reqLogger.Info(
+			"deleting backing PersistentVolume",
+			"PersistentVolume.Name", pv.Name,
+		)
+		self.recorder.Event(
+			volume, corev1.EventTypeNormal, "PvDeletion",
+			"backing PersistentVolume deleted",
+		)
+		return requeue(nil)
 	}
 
-	// PV already exists - don't requeue
-	reqLogger.Info("Skip reconcile: PersistentVolume already exists", "PersistentVolume.Name", found.Name)
+	// If we don't have a PV or it's only used by us we can reclaim the storage.
+	if pv == nil || isPersistentVolumeUnused(request, pv) {
+		return self.reclaimStorage(ctx, request, volume, pv)
+	}
 
-	return reconcile.Result{}, nil
+	// PersistentVolume still in use: wait before reclaiming the storage.
+	return delayedRequeue(nil)
 }
 
-func newPersistentVolumeForCR(cr *storagev1alpha1.Volume) *corev1.PersistentVolume {
+// Build a PersistentVolume from a Volume object.
+//
+// Arguments
+//     volume:      a Volume object
+//     volume_size: the volume size
+//
+// Returns
+//     The PersistentVolume representing the given Volume.
+func newPersistentVolume(
+	volume *storagev1alpha1.Volume, volume_size resource.Quantity,
+) *corev1.PersistentVolume {
 	return &corev1.PersistentVolume{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:   cr.Name,
-			Labels: map[string]string{},
-			Finalizers: []string{
-				"storage.metalk8s.scality.com/volume-protection",
-			},
+			Name:       volume.Name,
+			Labels:     map[string]string{},
+			Finalizers: []string{VOLUME_PROTECTION},
 		},
 		Spec: corev1.PersistentVolumeSpec{
```

Comment:
Need to copy `MountOptions` from the `StorageClass`.

Reply:
Done